### PR TITLE
NYPL Axis Service Tests

### PR DESCRIPTION
**Pull request by singhraman4282**
**What's this do?**
Adds tests for NYPLAxisService

**Why are we doing this? (w/ JIRA link if applicable)**
N/A

**How should this be tested? / Do these changes have associated tests?**
Cannot be QA tested

**Dependencies for merging? Releasing to production?**
N/A

**Does this include changes that require a new SimplyE build for QA?**
N/A

**Has the application documentation been updated for these changes?**
N/A

**Did someone actually run this code to verify it works?**
Raman Singh verified it

### CODE REVIEW

File: SimplifiedTests/MockServiceUsingAxisItemDownloader.swift [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+//  OpenEbooksTests
+//
+//  Created by Raman Singh on 2021-05-17.
+//  Copyright © 2021 NYPL Labs. All rights reserved.
+//
+
+import Foundation
+@testable import SimplyE
+
+class MockServiceUsingAxisItemDownloader {
```
Comment (ErnestFan) [inline SimplifiedTests/MockServiceUsingAxisItemDownloader.swift:12]: I think the name of this class is kinda confusing.
Can it be named something like `NYPLAxisItemDownloaderMockResponseService / NYPLAxisItemDownloaderMockSucceedService`? Or anything describe this class better.
@ettore any thoughts?

Comment (ettore) [reply SimplifiedTests/MockServiceUsingAxisItemDownloader.swift:12]: I'm not sure... I don't think this is mocking a service (it's not implementing any protocols), it just provides functionality to other mocks as part of their own class hierarchy. Mock ideally should be as simple as possible to avoid getting into situations where the flow of the test is hard to follow.

Maybe I would call it NYPLDownloadRunnerMock since it's triggering the downloader? but I'm actually not sure how this works to make a good suggestion.

Comment (singhraman4282) [reply SimplifiedTests/MockServiceUsingAxisItemDownloader.swift:12]: NYPLDownloadRunnerMock makes more sense. This class acts as a base class for NYPLAxisMetadataServiceMock,  NYPLAxisPackageServiceMock, & NYPLAxisLicenseServiceMock. If you believe this complicates the code I'll be more than happy to make changes to it.

**Review by ErnestFan — APPROVED** [review]

File: SimplifiedTests/MockServiceUsingAxisItemDownloader.swift (hunk 1 above)
Comment (ettore) [reply SimplifiedTests/MockServiceUsingAxisItemDownloader.swift]: generally speaking in tests is totally fine to repeat the same code over and over. The objective is to write tests that are easy to understand -- we don't want to look for logic bugs in tests, essentially. I do get it that like in this case there is some test logic that is repeated.

But no I'm not asking you to change it because I don't have better suggestions right now (I would have to look at it more deeply and I can't right now).

For the future though I would limit code reuse to things like helper functions but not much else. I'd try to avoid any logic like conditional branches, loops, subclassing-- not always possible but something to strive for.

File: SimplifiedTests/NYPLAxisTests/NYPLAxisServiceTests.swift [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+                                        packageService: succeedingPackageService)
+    
+    let task = URLSessionDownloadTask()
+    
+    downloadBroadcaster.downloadFailed = {
+      downloadFailedExpectation.fulfill()
+    }
+    
+    downloadBroadcaster.downloadSuccessful = {
+      downloadSucceededExpectation.fulfill()
```
Comment (ettore) [inline SimplifiedTests/NYPLAxisTests/NYPLAxisServiceTests.swift:53]: wouldn't this mean that we have to wait 10 seconds for this inverted expectation to be... not fulfilled? I haven't used inverted expectations often. 

wouldn't it be simpler to just do a XCFail() inside the success block instead? Same for the other tests, wherever we have explicit fail / success blocks.

File: SimplifiedTests/NYPLAxisTests/NYPLAxisServiceTests.swift [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+    }
+    
+    axisService.fulfillAxisLicense(downloadTask: task)
+    
+    wait(
+      for: [downloadFailedExpectation, downloadSucceededExpectation],
+      timeout: 10.0)
```
Comment (ettore) [inline SimplifiedTests/NYPLAxisTests/NYPLAxisServiceTests.swift:60]: just wondering if these can be reduced to 2-3 secs

File: SimplifiedTests/NYPLAxisTests/NYPLAxisServiceTests.swift [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+    let axisService = createAxisService(licenseService: succeedingLicenseService,
+                                        metadataService: succeedingMetadataService,
+                                        packageService: succeedingPackageService)
+    
+    let task = URLSessionDownloadTask()
+    
+    downloadBroadcaster.downloadSuccessful = {
+      downloadSucceededExpectation.fulfill()
+    }
+
```
Comment (ettore) [inline SimplifiedTests/NYPLAxisTests/NYPLAxisServiceTests.swift:134]: can we have a  `downloadFailed = { XCTFail() }` so we can fail faster?

File: SimplifiedTests/NYPLAxisTests/NYPLAxisServiceTests.swift (hunk 2 above)
Comment (singhraman4282) [reply SimplifiedTests/NYPLAxisTests/NYPLAxisServiceTests.swift:53]: Agreed. Fixed.

File: SimplifiedTests/NYPLAxisTests/NYPLAxisServiceTests.swift (hunk 3 above)
Comment (singhraman4282) [reply SimplifiedTests/NYPLAxisTests/NYPLAxisServiceTests.swift:60]: Done

File: SimplifiedTests/NYPLAxisTests/NYPLAxisServiceTests.swift (hunk 4 above)
Comment (singhraman4282) [reply SimplifiedTests/NYPLAxisTests/NYPLAxisServiceTests.swift:134]: Dine

Comment (ettore) [reply SimplifiedTests/NYPLAxisTests/NYPLAxisServiceTests.swift:116]: I don't see the downloadFailed callback anymore, maybe it got lost while rebasing?

**Review by ettore — APPROVED** [review]

File: SimplifiedTests/NYPLAxisTests/NYPLAxisServiceTests.swift (hunk 4 above)
Comment (singhraman4282) [reply SimplifiedTests/NYPLAxisTests/NYPLAxisServiceTests.swift:116]: rebases 😬

Comment (ettore) [reply SimplifiedTests/NYPLAxisTests/NYPLAxisServiceTests.swift:116]: I see it now, looks good 👍
